proposals/text/5629-file-similarity-retriever.md

```diff
@@ -0,0 +1,135 @@
+        inputs: [File]
+
+      - name: EmbeddingRetriever
+        inputs: [TextConverter]
+
+      - name: DocumentStore
+        inputs: [EmbeddingRetriever]
+
+```
+
+# Motivation
+
+Initially this retriever was implemented by Mathis for a customer case to quickly retrieve similar files given a file as input. The reason we would like to add it to Haystack is because we see wider use for this node in future customer and community cases.
+
+One example where this could be useful is in academic writing, if you have a large number of scientific sources (journals, books, conference proceedings) stored and you'd like to find all that are similar to a specific article or to what you've written so far (in the latter case, it can also be helpful for plagiarism detection). And as for industry use cases, report writing in any field can be facilitated by file similarity retrieval. It can also be very helpful for lawyers preparing a case as well as journalists doing research for their articles.
+
+# Detailed design
+
+You can see the code for this proposal here: https://github.com/deepset-ai/haystack/pull/5666/files.
+
+As a general description, the FileSimilarityRetriever works by:
+1. Getting all documents corresponding to the provided file ID from the document store. The documents can be obtained together with their embeddings, to save resources by avoiding re-calculating them at query time.
+2. Retrieving similar docs to each document from the file, using one or two chosen retrieval methods.
+3.  a) Returning the top-k retrieved similar documents, if only one retrieval method was used.
+    b) Returning an aggregated list of similar documents retrieved by both retrievers, if two retrieval methods were chosen. The results are aggregated based on the reciprocal rank fusion score, though it is also possible to keep the original score (if one of the retrievers assigned the document a relevance score) in the metadata.
+    Reciprocal rank fusion is calculated as 1 / (k + index)
+    k is set to 61 in reciprocal rank fusion based on a University of Waterloo paper (co-authored with Google) called "Reciprocal Rank Fusion outperforms Condorcet and individual Rank Learning Methods" [https://plg.uwaterloo.ca/~gvcormac/cormacksigir09-rrf.pdf] where k=60 was suggested, and 1 was added as python lists are 0-based and the paper used 1-based ranking.
+4. Returning the top-k similar files, based on an aggregated relevance score of all documents pertaining to the file. In the results the user sees the most relevant document from each file.
+We can also provide the user a choice to have the values of the metadata key used for aggregation (e.g. file IDs) returned instead of the most relevant documents, e.g. via an additional optional param "output" set by the user to "top_document" or "file_aggregation_key", with "top_document" being default since it looks more similar to the output of all pipelines (as haystack pipelines usually return natural language responses).
```

Comment:
@sjrl @mathislucka Great suggestion about letting the user choose what they want the output to be - I've added a line about that in the description 👍